### PR TITLE
Fix taxon count db loading

### PR DESCRIPTION
**Pull request by yunfangjuan**
# Description

Fixing constant db loading error. (caused when the records are already loaded.) 
https://jira.czi.team/browse/IDSEQ-471

## Type of change

- [X ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix a feature that knowingly causes existing functionality to not work as expected)

# Test and Reproduce
Outline the steps to test or reproduce the PR here.

1. 

## Impacted Areas in Application
List general components of the application that this PR will affect:

Which pages?

- [ ] Login Screen
- [ ] Single Upload Page
- [ ] Batch Upload Page
- [ ] All Projects Page
- [ ] Sample Details Page
- [ ] Report Page

Specific components 

* 

# Testing Script:

###Login Page
* Successfully login to the application and get taken to the project page

###Single Upload
* Using an AWS path that doesn't exist throws an error
* Succesfully uploaded files from a folder (use test host)
* Uploading a duplicate file throws an error
* Pipeline gets initiated successfully 

###Batch Upload
* Using an AWS path that doesn't exist throws an error
* Succesfully uploaded files from a folder (use test host)
* Navigate to batch upload settings page after upload 
* Host selection from previous page still selected 
* Uploading a duplicate file throws a warning
* Pipeline gets initiated successfully 

###All Project Page
* Switch between projects 
* Search for a sample
* Navigate through different pages of samples
* Select Upload new sample button
* Select sample to see generated report

###Sample Details Page
* Information is consistent from Project Page
* Notes are editable and save successfully 
* Can navigate to report page 

###Report Page
* Data loads succesfully 
* Collapse and Uncollapse button works
* Clicking on the taxonomy name gives you total reads for that taxon
* Category filter works
* Genus Search works
* Filters work as expected
* Disabling Filters works
* Filtering by column works 

# Checklist:

- [ ] I have run through the testing script to make sure current functionality is unchanged
- [ ] I have done relevant tests that prove my fix is effective or that my feature works
- [ ] I have spent time testing out edge cases for my feature
- [ ] I have updated the test script or pull request template if necessary
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/models/pipeline_run.rb [hunk 1]
```diff
@@ -723,7 +729,7 @@ def generate_aggregate_counts(tax_level_name)
     uncategorizable_id = TaxonLineage::MISSING_LINEAGE_ID.fetch(tax_level_name.to_sym, -9999)
     uncategorizable_name = "Uncategorizable as a #{tax_level_name}"
     TaxonCount.connection.execute(
-      "INSERT INTO taxon_counts(pipeline_run_id, tax_id, name,
+      "REPLACE INTO taxon_counts(pipeline_run_id, tax_id, name,
```
Comment (jameshowardwang) [inline app/models/pipeline_run.rb]: do you want REPLACE INTO or INSERT IGNORE INTO? are these always the exact same rows or are they sometimes actually being updated?

Comment (cdebourcy) [inline app/models/pipeline_run.rb]: Trying to understand better... wouldn't the block you added above remove the need for "replace"?

File: app/models/pipeline_run.rb [hunk 2]
```diff
@@ -335,6 +335,12 @@ def load_taxons(downloaded_json_path, refined = false)
     pipeline_output_dict = json_dict['pipeline_output']
     pipeline_output_dict.slice!('taxon_counts_attributes')
 
+    # check if there's any record loaded into taxon_counts. If so, skip
+    check_count_type = refined ? 'NT+' : 'NT'
+    loaded_records = TaxonCount.where(pipeline_run_id: id)
+                               .where(count_type: check_count_type).count
+    return if loaded_records > 0
```
Comment (jameshowardwang) [inline app/models/pipeline_run.rb]: is this the right thing to do? do all the records get loaded in one transaction? or is it possible that some taxons are loaded but not all of them?

File: app/models/pipeline_run.rb (hunk 1 above)
Comment (yunfangjuan) [reply app/models/pipeline_run.rb]: yes. i think so because we have NT and NT+ now.

File: app/models/pipeline_run.rb (hunk 2 above)
Comment (yunfangjuan) [reply app/models/pipeline_run.rb]: they are loaded in one transaction because how update_attributes work

File: app/models/pipeline_run.rb (hunk 1 above)
Comment (yunfangjuan) [reply app/models/pipeline_run.rb]: because we have NT/NT+ we kinda need this.

Comment (cdebourcy) [reply app/models/pipeline_run.rb]: Isn't it handled by "count_type"?

Comment (yunfangjuan) [reply app/models/pipeline_run.rb]: no. because this function will be called twice and the second time NT/NR records are already there, this will cause an insert error.

Comment (cdebourcy) [reply app/models/pipeline_run.rb]: ok
